google/cloud/pubsub/internal/streaming_subscription_batch_source.cc

```diff
@@ -316,34 +322,52 @@ void StreamingSubscriptionBatchSource::OnFinish(Status status) {
 
 void StreamingSubscriptionBatchSource::DrainQueues(
     std::unique_lock<std::mutex> lk, bool force_flush) {
-  auto const threshold = force_flush ? 1 : ack_batching_config_.max_batch_size;
-  if (ack_queue_.size() < threshold && nack_queue_.size() < threshold &&
-      deadlines_queue_.empty()) {
+  auto const threshold = ack_batching_config_.max_batch_size;
+  if (!force_flush && ack_queue_.size() < threshold &&
+      nack_queue_.size() < threshold) {
     return;
   }
   if (stream_state_ != StreamState::kActive || pending_write_) return;
+
+  auto const now = timer_config_.clock();
+  flush_deadline_ = now + ack_batching_config_.max_hold_time;
+
   auto stream = stream_;
   pending_write_ = true;
+  google::pubsub::v1::StreamingPullRequest request;
 
   std::vector<std::string> acks;
   acks.swap(ack_queue_);
   std::vector<std::string> nacks;
   nacks.swap(nack_queue_);
-  std::vector<std::pair<std::string, std::chrono::seconds>> deadlines;
-  deadlines.swap(deadlines_queue_);
+
+  std::vector<std::string> lease_extensions;
+  // We try to avoid work while holding the `lk` lock, but rarely (about once
+  // every 5 seconds) we need to update the leases. That requires a potentially
+  // long loop.
```

Comment:
s/loop/time/ perhaps?

Reply:
Fixed. And fixed other things as I was looking at this code again.